persist/fs/commitlog/reader.go

```diff
@@ -109,58 +192,267 @@ func (r *reader) Read() (
-		decoded, err := r.metadataDecoder.DecodeLogMetadata()
-		if err != nil {
-			resultErr = err
+func (r *reader) startBackgroundWorkers() error {
+	// Make sure background workers are never setup more than once
+	set := atomic.CompareAndSwapInt64(&r.bgWorkersInitialized, 0, 1)
+	if !set {
+		return errCommitLogReaderMultipleReadloops
+	}
+
+	// Start background worker goroutines
+	go r.readLoop()
+	for i, decoderBuf := range r.decoderBufs {
+		localDecoderBuf, outBuf := decoderBuf, r.outBufs[i]
+		go r.decoderLoop(localDecoderBuf, outBuf)
+	}
+
+	return nil
+}
+
+func (r *reader) readLoop() {
+	index := 0
+	defer r.shutdown()
+	for {
+		select {
+		case <-r.cancelCtx.Done():
 			return
+		default:
+			data, err := r.readChunk()
+			// Distribute the decoding work in round-robin fashion so that when we
+			// read round-robin, we get the data back in the original order.
+			r.decoderBufs[index%r.numConc] <- decoderArg{
+				bytes: data,
+				err:   err,
+			}
+			index++
+			if err == io.EOF {
+				return
+			}
+		}
+	}
+}
+
+func (r *reader) shutdown() {
+	for _, decoderBuf := range r.decoderBufs {
+		close(decoderBuf)
+	}
+	r.shutdownCh <- r.close()
+}
+
+func (r *reader) decoderLoop(inBuf <-chan decoderArg, outBuf chan<- readResponse) {
+	decodingOpts := r.opts.FilesystemOptions().DecodingOptions()
+	decoder := msgpack.NewDecoder(decodingOpts)
+	metadataDecoder := msgpack.NewDecoder(decodingOpts)
+
+	for arg := range inBuf {
+		readResponse := readResponse{}
+		if arg.err != nil {
+			readResponse.resultErr = arg.err
+			r.writeToOutBuf(outBuf, readResponse)
+			continue
 		}
 
-		id := r.bytesPool.Get(len(decoded.ID))
-		id.IncRef()
-		defer id.DecRef()
-		id.AppendAll(decoded.ID)
-
-		namespace := r.bytesPool.Get(len(decoded.Namespace))
-		namespace.IncRef()
-		defer namespace.DecRef()
-		namespace.AppendAll(decoded.Namespace)
-
-		metadata = Series{
-			UniqueIndex: entry.Index,
-			ID:          ts.BinaryID(id),
-			Namespace:   ts.BinaryID(namespace),
-			Shard:       decoded.Shard,
+		arg.bytes.IncRef()
+		decoder.Reset(arg.bytes.Get())
+		entry, err := decoder.DecodeLogEntry()
+		if err != nil {
+			readResponse.resultErr = err
+			r.writeToOutBuf(outBuf, readResponse)
+			continue
 		}
-		ok = true
 
-		r.metadataLookup[entry.Index] = metadata
+		if len(entry.Metadata) != 0 {
+			metadataDecoder.Reset(entry.Metadata)
+			decoded, err := metadataDecoder.DecodeLogMetadata()
+			if err != nil {
+				readResponse.resultErr = err
+				r.writeToOutBuf(outBuf, readResponse)
+				continue
+			}
+
+			id := r.bytesPool.Get(len(decoded.ID))
+			id.IncRef()
+			id.AppendAll(decoded.ID)
+
+			namespace := r.bytesPool.Get(len(decoded.Namespace))
+			namespace.IncRef()
+			namespace.AppendAll(decoded.Namespace)
+
+			r.metadata.Lock()
+			_, ok := r.metadata.lookup[entry.Index]
+			// If the metadata already exists, we can skip this step
+			if ok {
+				id.DecRef()
+				id.Finalize()
+				namespace.DecRef()
+				namespace.Finalize()
+			} else {
```

Comment:
could you `continue` in the if branch to avoid this else?

Reply:
I don't think so for the same reason as above, I need to execute the code after the if statement regardless